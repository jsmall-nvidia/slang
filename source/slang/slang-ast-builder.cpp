// slang-ast-builder.cpp
#include "slang-ast-builder.h"
#include <assert.h>

#include "slang-compiler.h"

namespace Slang {

ASTBuilder::ASTBuilder():
    m_session(nullptr)
{
    memset(m_builtinTypes, 0, sizeof(m_builtinTypes));
}

void ASTBuilder::init(Session* session)
{
    SLANG_ASSERT(session);

    for (Index i = 0; i < Index(BaseType::CountOf); ++i)
    {
        m_builtinTypes[i] = session->getBuiltinType(BaseType(i));
    }

    m_errorType = session->getErrorType();
    m_initializerListType = session->getInitializerListType();
    m_overloadedType = session->getOverloadedType();

    //m_constExprRate = session->g
    //m_irBasicBlockType = session->get

    m_stringType = session->getStringType();
    m_enumTypeType = session->getEnumTypeType();
}

RefPtr<PtrType> ASTBuilder::getPtrType( RefPtr<Type>    valueType)
{
    return getPtrType(valueType, "PtrType").dynamicCast<PtrType>();
}

// Construct the type `Out<valueType>`
RefPtr<OutType> ASTBuilder::getOutType(RefPtr<Type> valueType)
{
    return getPtrType(valueType, "OutType").dynamicCast<OutType>();
}

RefPtr<InOutType> ASTBuilder::getInOutType(RefPtr<Type> valueType)
{
    return getPtrType(valueType, "InOutType").dynamicCast<InOutType>();
}

RefPtr<RefType> ASTBuilder::getRefType(RefPtr<Type> valueType)
{
    return getPtrType(valueType, "RefType").dynamicCast<RefType>();
}

RefPtr<PtrTypeBase> ASTBuilder::getPtrType(RefPtr<Type> valueType, char const* ptrTypeName)
{
    auto genericDecl = findMagicDecl(m_session, ptrTypeName).dynamicCast<GenericDecl>();
    return getPtrType(valueType, genericDecl);
}

RefPtr<PtrTypeBase> ASTBuilder::getPtrType(RefPtr<Type> valueType, GenericDecl* genericDecl)
{
    auto typeDecl = genericDecl->inner;

    auto substitutions = create<GenericSubstitution>();
    substitutions->genericDecl = genericDecl;
    substitutions->args.add(valueType);

    auto declRef = DeclRef<Decl>(typeDecl.Ptr(), substitutions);
    auto rsType = DeclRefType::create(this, declRef);
    return as<PtrTypeBase>(rsType);
}

RefPtr<ArrayExpressionType> ASTBuilder::getArrayType(Type* elementType, IntVal* elementCount)
{
    RefPtr<ArrayExpressionType> arrayType = create<ArrayExpressionType>();
    arrayType->baseType = elementType;
    arrayType->arrayLength = elementCount;
    return arrayType;
}

RefPtr<VectorExpressionType> ASTBuilder::getVectorType(
    RefPtr<Type>    elementType,
    RefPtr<IntVal>  elementCount)
{
    auto vectorGenericDecl = findMagicDecl(m_session, "Vector").as<GenericDecl>();
        
    auto vectorTypeDecl = vectorGenericDecl->inner;

    auto substitutions = new GenericSubstitution();
    substitutions->genericDecl = vectorGenericDecl.Ptr();
    substitutions->args.add(elementType);
    substitutions->args.add(elementCount);

    auto declRef = DeclRef<Decl>(vectorTypeDecl.Ptr(), substitutions);

    return DeclRefType::create(this, declRef).as<VectorExpressionType>();
}

RefPtr<TypeType> ASTBuilder::getTypeType(Type* type)
{
    return create<TypeType>(type);
}

} // namespace Slang
