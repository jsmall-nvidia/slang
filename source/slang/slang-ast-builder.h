// slang-ast-dump.h
#ifndef SLANG_AST_BUILDER_H
#define SLANG_AST_BUILDER_H

#include "slang-ast-support-types.h"
#include "slang-ast-all.h"

#include "../core/slang-type-traits.h"

namespace Slang
{

class SharedASTBuilder : public RefObject
{
    friend class ASTBuilder;
public:
    
    void registerBuiltinDecl(RefPtr<Decl> decl, RefPtr<BuiltinTypeModifier> modifier);
    void registerMagicDecl(RefPtr<Decl> decl, RefPtr<MagicTypeModifier> modifier);

        /// Get the string type
    Type* getStringType();
        /// Get the enum type type
    Type* getEnumTypeType();

    const ReflectClassInfo* findClassInfo(Name* name);
    SyntaxClass<RefObject> findSyntaxClass(Name* name);

    const ReflectClassInfo* findClassInfo(const UnownedStringSlice& slice);
    SyntaxClass<RefObject> findSyntaxClass(const UnownedStringSlice& slice);

        // Look up a magic declaration by its name
    RefPtr<Decl> findMagicDecl(String const& name);

        /// A name pool that can be used for lookup for findClassInfo etc. It is the same pool as the Session.
    NamePool* getNamePool() { return m_namePool; }

        /// Must be called before used
    void init(Session* session);

    SharedASTBuilder();

    ~SharedASTBuilder();

protected:
    // State shared between ASTBuilders

    RefPtr<Type> m_errorType;
    RefPtr<Type> m_initializerListType;
    RefPtr<Type> m_overloadedType;

    // The following types are created lazily, such that part of their definition
    // can be in the standard library
    // 
    // Note(tfoley): These logically belong to `Type`,
    // but order-of-declaration stuff makes that tricky
    //
    // TODO(tfoley): These should really belong to the compilation context!
    //
    RefPtr<Type> m_stringType;
    RefPtr<Type> m_enumTypeType;

    RefPtr<Type> m_builtinTypes[Index(BaseType::CountOf)];

    Dictionary<String, Decl*> m_magicDecls;

    Dictionary<UnownedStringSlice, const ReflectClassInfo*> m_sliceToTypeMap;
    Dictionary<Name*, const ReflectClassInfo*> m_nameToTypeMap;
    
    NamePool* m_namePool = nullptr;

    // This is a private builder used for these shared types
    ASTBuilder* m_astBuilder = nullptr;
    Session* m_session = nullptr;
};

class ASTBuilder : public RefObject
{
    friend class SharedASTBuilder;
public:

    // For compile time check to see if thing being constructed is an AST type
    template <typename T>
    struct IsValidType
    {
        enum
        {
            Value = IsBaseOf<NodeBase, T>::Value
        };
    };

        /// Create AST type. 
    template <typename T>
    T* create() { SLANG_COMPILE_TIME_ASSERT(IsValidType<T>::Value);  T* node = new T; node->init(T::kType, this); return node; }

    template<typename T, typename P0>
    T* create(const P0& p0) { SLANG_COMPILE_TIME_ASSERT(IsValidType<T>::Value); T* node = new T(p0); node->init(T::kType, this); return node;}

    template<typename T, typename P0, typename P1>
    T* create(const P0& p0, const P1& p1) { SLANG_COMPILE_TIME_ASSERT(IsValidType<T>::Value); T* node = new T(p0, p1); node->init(T::kType, this); return node; }

        /// Get the built in types
    SLANG_FORCE_INLINE Type* getBoolType() { return m_sharedASTBuilder->m_builtinTypes[Index(BaseType::Bool)]; }
    SLANG_FORCE_INLINE Type* getHalfType() { return m_sharedASTBuilder->m_builtinTypes[Index(BaseType::Half)]; }
    SLANG_FORCE_INLINE Type* getFloatType() { return m_sharedASTBuilder->m_builtinTypes[Index(BaseType::Float)]; }
    SLANG_FORCE_INLINE Type* getDoubleType() { return m_sharedASTBuilder->m_builtinTypes[Index(BaseType::Double)]; }
    SLANG_FORCE_INLINE Type* getIntType() { return m_sharedASTBuilder->m_builtinTypes[Index(BaseType::Int)]; }
    SLANG_FORCE_INLINE Type* getInt64Type() { return m_sharedASTBuilder->m_builtinTypes[Index(BaseType::Int64)]; }
    SLANG_FORCE_INLINE Type* getUIntType() { return m_sharedASTBuilder->m_builtinTypes[Index(BaseType::UInt)]; }
    SLANG_FORCE_INLINE Type* getUInt64Type() { return m_sharedASTBuilder->m_builtinTypes[Index(BaseType::UInt64)]; }
    SLANG_FORCE_INLINE Type* getVoidType() { return m_sharedASTBuilder->m_builtinTypes[Index(BaseType::Void)]; }

        /// Get a builtin type by the BaseType
    SLANG_FORCE_INLINE Type* getBuiltinType(BaseType flavor) { return m_sharedASTBuilder->m_builtinTypes[Index(flavor)]; }

    Type* getInitializerListType() { return m_sharedASTBuilder->m_initializerListType; }
    Type* getOverloadedType() { return m_sharedASTBuilder->m_overloadedType; }
    Type* getErrorType() { return m_sharedASTBuilder->m_errorType; }
    Type* getStringType() { return m_sharedASTBuilder->getStringType(); }
    Type* getEnumTypeType() { return m_sharedASTBuilder->getEnumTypeType(); }

        // Construct the type `Ptr<valueType>`, where `Ptr`
        // is looked up as a builtin type.
    RefPtr<PtrType> getPtrType(RefPtr<Type> valueType);

        // Construct the type `Out<valueType>`
    RefPtr<OutType> getOutType(RefPtr<Type> valueType);

        // Construct the type `InOut<valueType>`
    RefPtr<InOutType> getInOutType(RefPtr<Type> valueType);

        // Construct the type `Ref<valueType>`
    RefPtr<RefType> getRefType(RefPtr<Type> valueType);

        // Construct a pointer type like `Ptr<valueType>`, but where
        // the actual type name for the pointer type is given by `ptrTypeName`
    RefPtr<PtrTypeBase> getPtrType(RefPtr<Type> valueType, char const* ptrTypeName);

        // Construct a pointer type like `Ptr<valueType>`, but where
        // the generic declaration for the pointer type is `genericDecl`
    RefPtr<PtrTypeBase> getPtrType(RefPtr<Type> valueType, GenericDecl* genericDecl);

    RefPtr<ArrayExpressionType> getArrayType(Type* elementType, IntVal* elementCount);

    RefPtr<VectorExpressionType> getVectorType(RefPtr<Type> elementType, RefPtr<IntVal> elementCount);

    RefPtr<TypeType> getTypeType(Type* type);

        /// Helpers to get type info from the SharedASTBuilder
    const ReflectClassInfo* findClassInfo(const UnownedStringSlice& slice) { return m_sharedASTBuilder->findClassInfo(slice); }
    SyntaxClass<RefObject> findSyntaxClass(const UnownedStringSlice& slice) { return m_sharedASTBuilder->findSyntaxClass(slice); }

    const ReflectClassInfo* findClassInfo(Name* name) { return m_sharedASTBuilder->findClassInfo(name); }
    SyntaxClass<RefObject> findSyntaxClass(Name* name) { return m_sharedASTBuilder->findSyntaxClass(name); }

        /// Get the shared AST builder
    SharedASTBuilder* getSharedASTBuilder() { return m_sharedASTBuilder; }

        /// Get the global session
    Session* getGlobalSession() { return m_sharedASTBuilder->m_session; }

        /// Ctor
    ASTBuilder(SharedASTBuilder* sharedASTBuilder);

protected:
    // Special default Ctor that can only be used by SharedASTBuilder
    ASTBuilder();

    SharedASTBuilder* m_sharedASTBuilder;
};

} // namespace Slang

#endif
