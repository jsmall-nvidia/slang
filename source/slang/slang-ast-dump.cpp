// slang-ast-dump.cpp
#include "slang-ast-dump.h"
#include <assert.h>

#include "slang-compiler.h"

#include "../core/slang-string.h"

#include "slang-ast-generated-macro.h"

// There could be an argument for using the SourceWriter, but whilst that has some useful stuff
// like the automatic indentation, it is a little clumsy in some other ways, like how to build up
// a result via << operators is not available.
namespace Slang {

namespace { // anonymous

struct Context
{
    struct ObjectInfo
    {
        const ReflectClassInfo* m_typeInfo;
        RefObject* m_object;
        bool m_isDumped;
    };

    struct ScopeWrite
    {
        ScopeWrite(Context* context):
            m_context(context)
        {
            if (m_context->m_scopeWriteCount == 0)
            {
                m_context->m_buf.Clear();
            }
            m_context->m_scopeWriteCount++;
        }

        ~ScopeWrite()
        {
            if (--m_context->m_scopeWriteCount == 0)
            {
                m_context->m_writer->emit(m_context->m_buf);
            }
        }

        StringBuilder& getBuf() { return m_context->m_buf; }

        operator StringBuilder&() { return m_context->m_buf; }

        Context* m_context;
    };

    void dumpObject(const ReflectClassInfo& type, RefObject* obj);

    void dumpObjectFull(const ReflectClassInfo& type, RefObject* obj, Index objIndex);
    void dumpObjectReference(const ReflectClassInfo& type, RefObject* obj, Index objIndex);

    void dump(NodeBase* node)
    {
        if (node == nullptr)
        {
            dumpPtr(nullptr);
        }
        else
        {
            dumpObject(node->getClassInfo(), node);
        }
    }

    void dump(Substitutions* subs)
    {
        if (subs == nullptr)
        {
            dumpPtr(nullptr);
        }
        else
        {
            dumpObject(subs->getClassInfo(), subs);
        }
    }

    void dump(const Name* name)
    {
        if (name == nullptr)
        {
            dumpPtr(nullptr);
        }
        else
        {
            dump(name->text);
        }
    }

    void dump(const RefObject* obj)
    {
        if (obj == nullptr)
        {
            dumpPtr(nullptr);
        }
        else
        {
            // We don't know what this is!
            ScopeWrite(this).getBuf() << "Unknown@" << size_t(obj);
        }
    }

    template <typename T>
    void dump(const List<T>& list)
    {
        m_writer->emit(" { \n");
        m_writer->indent();
        for (Index i = 0; i < list.getCount(); ++i)
        {
            dump(list[i]);
            if (i < list.getCount() - 1)
            {
                m_writer->emit(",\n");
            }
            else
            {
                m_writer->emit("\n");
            }
        }
        m_writer->dedent();
        m_writer->emit("}");
    }

    void dump(SourceLoc sourceLoc)
    {
        SourceManager* manager = m_writer->getSourceManager();

        {
            ScopeWrite(this).getBuf() << "SourceLoc(" << sourceLoc.getRaw() << ")";
        }

        if (manager && sourceLoc.isValid())
        {
            HumaneSourceLoc humaneLoc = manager->getHumaneLoc(sourceLoc);
            ScopeWrite(this).getBuf() << " " << humaneLoc.pathInfo.foundPath << ":" << humaneLoc.line;   
        }
    }

    static const char _getHexDigit(UInt v)
    {
        return (v < 10) ? char(v + '0') : char('a' + v - 10);
    }

    void dump(const UnownedStringSlice& slice)
    {
        m_writer->emitChar('\"');

        {
            ScopeWrite scope(this);
            auto& buf = scope.getBuf();
            for (const char c : slice)
            {
                if (c < 0x20 || c >= 0x80)
                {
                    buf << "\\0x" <<  _getHexDigit(UInt32(c) >> 4) << _getHexDigit(c & 0xf);
                }
                else
                {
                    buf << c;
                }
            }
        }
        m_writer->emitChar('\"');
    }
    
    void dump(const Token& token)
    {
        ScopeWrite(this).getBuf() << " { " << TokenTypeToString(token.type) << ", ";
        dump(token.loc);
        dump(token.Content);
        m_writer->emit(" }");
    }

    Index getObjectIndex(const ReflectClassInfo& typeInfo, RefObject* obj)
    {
        Index* indexPtr = m_objectMap.TryGetValueOrAdd(obj, m_objects.getCount());
        if (indexPtr)
        {
            return *indexPtr;
        }

        ObjectInfo info;
        info.m_isDumped = false;
        info.m_object = obj;
        info.m_typeInfo = &typeInfo;

        m_objects.add(info);
        return m_objects.getCount() - 1;
    }

    void dump(uint32_t v)
    {
        m_writer->emit(UInt(v));
    }

    void dump(Index v)
    {
        m_writer->emit(v);
    }

    void dump(int32_t v)
    {
        m_writer->emit(v);
    }
    void dump(FloatingPointLiteralValue v)
    {
        m_writer->emit(v);
    }


    void dump(const SemanticVersion& version)
    {
        ScopeWrite(this).getBuf() << UInt(version.m_major) << "." << UInt(version.m_minor) << "." << UInt(version.m_patch);
    }
    void dump(const NameLoc& nameLoc)
    {
        m_writer->emit("NameLoc{");
        if (nameLoc.name)
        {
            dump(nameLoc.name->text.getUnownedSlice());
        }
        else
        {
            dumpPtr(nullptr);
        }
        m_writer->emit(", ");
        dump(nameLoc.loc);
        m_writer->emit(" }");
    }
    void dump(BaseType baseType)
    {
        m_writer->emit(BaseTypeInfo::asText(baseType));
    }
    void dump(Stage stage)
    {
        m_writer->emit(getStageName(stage));
    }
    void dump(ImageFormat imageFormat)
    {
        m_writer->emit(getGLSLNameForImageFormat(imageFormat));
    }

    void dump(const String& string)
    {
        dump(string.getUnownedSlice());
    }

    void dump(const DiagnosticInfo* info)
    {
        ScopeWrite(this).getBuf() << "DiagnosticInfo {" << info->id << "}";
    }
    void dump(const Layout* layout)
    {
        ScopeWrite(this).getBuf() << "Layout@" << size_t(layout); 
    }

    void dump(const Modifiers& modifiers)
    {
        auto& nonConstModifiers = const_cast<Modifiers&>(modifiers);

        m_writer->emit(" { \n");
        m_writer->indent();

        for (const auto& mod : nonConstModifiers)
        {
            dump(mod);
            m_writer->emit("\n");
        }

        m_writer->dedent();
        m_writer->emit("}");
    }

    template <typename T>
    void dump(const SyntaxClass<T>& cls)
    {
        m_writer->emit(cls.classInfo->m_name);
    }

    template <typename KEY, typename VALUE>
    void dump(const Dictionary<KEY, VALUE>& dict)
    {
        m_writer->emit(" { \n");
        m_writer->indent();

        for (auto iter : dict)
        {
            const auto& key = iter.Key;
            const auto& value = iter.Value;

            dump(key);
            m_writer->emit(" : ");
            dump(value);

            m_writer->emit("\n");
        }

        m_writer->dedent();
        m_writer->emit("}");
    }

    void dump(const DeclCheckStateExt& extState)
    {
        auto state = extState.getState();
      
        ScopeWrite(this).getBuf() << "DeclCheckStateExt{" << extState.isBeingChecked() << ", " << Index(state) << "}";
    }

    void dump(TextureFlavor texFlavor)
    {
        m_buf.Clear();
        m_buf << "TextureFlavor{" << Index(texFlavor.flavor) << "}";
        m_writer->emit(m_buf);
    }

    void dump(SamplerStateFlavor flavor)
    {
        switch (flavor)
        {
            case SamplerStateFlavor::SamplerState: m_writer->emit("sampler"); break;
            case SamplerStateFlavor::SamplerComparisonState: m_writer->emit("samplerComparison"); break;
            default: m_writer->emit("unknown"); break;
        }
    }

    void dump(const QualType& qualType)
    {
        if (qualType.IsLeftValue)
        {
            m_writer->emit("left ");
        }
        else
        {
            m_writer->emit("right ");
        }
        dump(qualType.type);
    }

    void dumpPtr(const void* ptr)
    {
        if (ptr)
        {
            ScopeWrite(this).getBuf() << "Unknown@" << size_t(ptr);
        }
        else
        {
            m_writer->emit("null");
        }
    }

    void dump(SyntaxParseCallback callback) { dumpPtr((const void*)callback); }

    template <typename T, int SIZE>
    void dump(const T (&in)[SIZE])
    {
        m_writer->emit(" { \n");
        m_writer->indent();

        for (Index i = 0; i < Index(SIZE); ++i)
        {
            dump(in[i]);
            if (i < Index(SIZE) - 1)
            {
                m_writer->emit(", ");
            }
            m_writer->emit("\n");
        }

        m_writer->dedent();
        m_writer->emit("}");
    }

    //void dump(const void* ptr) { dumpPtr(ptr); }

    void dump(const LookupResult& result)
    {
        auto& nonConstResult = const_cast<LookupResult&>(result);

        m_writer->emit(" { \n");
        m_writer->indent();

        for (auto item : nonConstResult)
        {
            // TODO(JS):
            m_writer->emit("...\n");
        }

        m_writer->dedent();
        m_writer->emit("}");
    }
    void dump(const GlobalGenericParamSubstitution::ConstraintArg& arg)
    {
        m_writer->emit(" { \n");
        m_writer->indent();

        dump(arg.decl);
        m_writer->emit(",\n");
        dump(arg.val);
        m_writer->emit("\n");

        m_writer->dedent();
        m_writer->emit("}");
    }
    void dump(const TypeExp& exp)
    {
        m_writer->emit(" { \n");
        m_writer->indent();

        dump(exp.exp);
        m_writer->emit(",\n");
        dump(exp.type);
        m_writer->emit("\n");

        m_writer->dedent();
        m_writer->emit("}");
    }
    void dump(const ExpandedSpecializationArg& arg)
    {
        dump(arg.witness);
    }

    void dump(const TransparentMemberInfo& memInfo)
    {
        dump(memInfo.decl);
    }

    void dumpRemaining()
    {
        // Have to keep checking count, as dumping objects can add objects
        for (Index i = 0; i < m_objects.getCount(); ++i)
        {
            ObjectInfo& info = m_objects[i];
            if (!info.m_isDumped)
            {
                dumpObjectFull(*info.m_typeInfo, info.m_object, i);
            }
        }
    }

    template <typename T>
    void dumpField(const char* name, const T& value)
    {
        m_writer->emit(name);
        m_writer->emit(" : ");
        dump(value);
        m_writer->emit("\n");
    }

    void dumpObjectFull(NodeBase* node);
    void dumpObjectFull(Substitutions* subs);

    Context(SourceWriter* writer, ASTDumpUtil::Style dumpStyle):
        m_writer(writer),
        m_scopeWriteCount(0),
        m_dumpStyle(dumpStyle)
    {
    }

    ASTDumpUtil::Style m_dumpStyle;

    Index m_scopeWriteCount;

    SourceWriter* m_writer;

    Dictionary<RefObject*, Index> m_objectMap;  ///< Object index
    List<ObjectInfo> m_objects;

    StringBuilder m_buf;
};

} // anonymous

// Lets generate functions one for each that attempts to write out *it's* fields.
// We can write out the Super types fields by looking that up

struct ASTDumpAccess
{
#define SLANG_AST_DUMP_FIELD(FIELD_NAME, TYPE, param) context.dumpField(#FIELD_NAME, node->FIELD_NAME); 

#define SLANG_AST_DUMP_FIELDS_IMPL(NAME, SUPER, ORIGIN, LAST, MARKER, TYPE, param) \
static void dumpFields_##NAME(NAME* node, Context& context) \
{ \
    SLANG_UNUSED(node); \
    SLANG_UNUSED(context); \
    SLANG_FIELDS_ASTNode_##NAME(SLANG_AST_DUMP_FIELD, _) \
}

SLANG_ALL_ASTNode_Substitutions(SLANG_AST_DUMP_FIELDS_IMPL, _)
SLANG_ALL_ASTNode_NodeBase(SLANG_AST_DUMP_FIELDS_IMPL, _)

};

#define SLANG_AST_GET_DUMP_FUNC(NAME, SUPER, ORIGIN, LAST, MARKER, TYPE, param) m_funcs[Index(ASTNodeType::NAME)] = (DumpFieldsFunc)&ASTDumpAccess::dumpFields_##NAME;

typedef void (*DumpFieldsFunc)(RefObject* obj, Context& context);

struct DumpFieldFuncs
{
    DumpFieldFuncs()
    {
        memset(m_funcs, 0, sizeof(m_funcs));
        SLANG_ALL_ASTNode_Substitutions(SLANG_AST_GET_DUMP_FUNC, _)
        SLANG_ALL_ASTNode_NodeBase(SLANG_AST_GET_DUMP_FUNC, _)
    }

    DumpFieldsFunc m_funcs[Index(ASTNodeType::CountOf)];
};

static const DumpFieldFuncs s_funcs;

void Context::dumpObjectReference(const ReflectClassInfo& type, RefObject* obj, Index objIndex)
{
    SLANG_UNUSED(obj);
    ScopeWrite(this).getBuf() << type.m_name << "@" << objIndex;
}

void Context::dumpObjectFull(const ReflectClassInfo& type, RefObject* obj, Index objIndex)
{
    ObjectInfo& info = m_objects[objIndex];
    SLANG_ASSERT(info.m_isDumped == false);
    info.m_isDumped = true;

    // We need to dump the fields.

    ScopeWrite(this).getBuf() << type.m_name << "(" << objIndex << ") {\n";
    m_writer->indent();

    List<const ReflectClassInfo*> allTypes;
    {
        const ReflectClassInfo* curType = &type;
        do
        {
            allTypes.add(curType);
            curType = curType->m_superClass;
        } while (curType);
    }

    // Okay we go backwards so we output in the 'normal' order
    for (Index i = allTypes.getCount() - 1; i >= 0; --i)
    {
        const ReflectClassInfo* curType = allTypes[i];
        DumpFieldsFunc func = s_funcs.m_funcs[Index(curType->m_classId)];
        if (func)
        {
            func(obj, *this);
        }
    }

    m_writer->dedent();
    m_writer->emit("}\n");
}

void Context::dumpObject(const ReflectClassInfo& typeInfo, RefObject* obj)
{
    Index index = getObjectIndex(typeInfo, obj);

    ObjectInfo& info = m_objects[index];
    if (info.m_isDumped || m_dumpStyle == ASTDumpUtil::Style::Flat)
    {
        dumpObjectReference(typeInfo, obj, index);
    }
    else
    {
        dumpObjectFull(typeInfo, obj, index);
    }
}

void Context::dumpObjectFull(NodeBase* node)
{
    if (!node)
    {
        dumpPtr(nullptr);
    }
    else
    {
        const ReflectClassInfo& typeInfo = node->getClassInfo();
        Index index = getObjectIndex(typeInfo, node);
        dumpObjectFull(typeInfo, node, index);
    }
}

void Context::dumpObjectFull(Substitutions* subs)
{
    if (!subs)
    {
        dumpPtr(nullptr);
    }
    else
    {
        const ReflectClassInfo& typeInfo = subs->getClassInfo();
        Index index = getObjectIndex(typeInfo, subs);
        dumpObjectFull(typeInfo, subs, index);
    }
}

/* static */void ASTDumpUtil::dump(NodeBase* node, Style style, SourceWriter* writer)
{
    Context context(writer, style);
    context.dumpObjectFull(node);
    context.dumpRemaining();
}

/* static */void ASTDumpUtil::dump(Substitutions* subs, Style style, SourceWriter* writer)
{
    Context context(writer, style);
    context.dumpObjectFull(subs);
    context.dumpRemaining();

}

} // namespace Slang
