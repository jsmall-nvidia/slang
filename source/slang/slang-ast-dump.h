// slang-ast-dump.h
#ifndef SLANG_AST_DUMP_H
#define SLANG_AST_DUMP_H

#include "slang-syntax.h"

#include "slang-emit-source-writer.h"

namespace Slang
{

struct ASTDumpAccess;

struct ASTDumpUtil
{
    enum class Style
    {
        Hierachical,
        Flat,
    };

    static void dump(NodeBase* node, Style style, SourceWriter* writer);
};

} // namespace Slang

#endif
