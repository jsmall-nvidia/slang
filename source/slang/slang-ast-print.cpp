// slang-ast-print.cpp
#include "slang-ast-print.h"

#include "slang-check-impl.h"

namespace Slang {

void ASTPrinter::addType(Type* type)
{
    m_builder << type->toString();
}

void ASTPrinter::addVal(Val* val)
{
    m_builder << val->toString();
}

void ASTPrinter::addDeclName(Decl* decl)
{
    if (as<ConstructorDecl>(decl))
    {
        m_builder << "init";
    }
    else if (as<SubscriptDecl>(decl))
    {
        m_builder << "subscript";
    }
    else
    {
        m_builder << getText(decl->getName());
    }
}

void ASTPrinter::addDeclPath(const DeclRef<Decl>& declRef)
{
    auto& sb = m_builder;

    // Find the parent declaration
    auto parentDeclRef = declRef.getParent();

    // If the immediate parent is a generic, then we probably
    // want the declaration above that...
    auto parentGenericDeclRef = parentDeclRef.as<GenericDecl>();
    if (parentGenericDeclRef)
    {
        parentDeclRef = parentGenericDeclRef.getParent();
    }

    // Depending on what the parent is, we may want to format things specially
    if (auto aggTypeDeclRef = parentDeclRef.as<AggTypeDecl>())
    {
        addDeclPath(aggTypeDeclRef);
        sb << ".";
    }

    addDeclName(declRef.getDecl());

    // If the parent declaration is a generic, then we need to print out its
    // signature
    if (parentGenericDeclRef)
    {
        auto genSubst = as<GenericSubstitution>(declRef.substitutions.substitutions);
        SLANG_RELEASE_ASSERT(genSubst);
        SLANG_RELEASE_ASSERT(genSubst->genericDecl == parentGenericDeclRef.getDecl());

        // If the name we printed previously was an operator
        // that ends with `<`, then immediately printing the
        // generic arguments inside `<...>` may cause it to
        // be hard to parse the operator name visually.
        //
        // We thus include a space between the declaration name
        // and its generic arguments in this case.
        //
        if (sb.endsWith("<"))
        {
            sb << " ";
        }

        sb << "<";
        bool first = true;
        for (auto arg : genSubst->args)
        {
            // When printing the representation of a specialized
            // generic declaration we don't want to include the
            // argument values for subtype witnesses since these
            // do not correspond to parameters of the generic
            // as the user sees it.
            //
            if (as<Witness>(arg))
                continue;

            if (!first) sb << ", ";
            addVal(arg);
            first = false;
        }
        sb << ">";
    }
}

void ASTPrinter::addDeclParams(const DeclRef<Decl>& declRef)
{
    auto& sb = m_builder;

    if (auto funcDeclRef = declRef.as<CallableDecl>())
    {
        // This is something callable, so we need to also print parameter types for overloading
        sb << "(";

        bool first = true;
        for (auto paramDeclRef : getParameters(funcDeclRef))
        {
            if (!first) sb << ", ";

            addType(getType(m_astBuilder, paramDeclRef));

            first = false;
        }

        sb << ")";
    }
    else if (auto genericDeclRef = declRef.as<GenericDecl>())
    {
        sb << "<";
        bool first = true;
        for (auto paramDeclRef : getMembers(genericDeclRef))
        {
            if (auto genericTypeParam = paramDeclRef.as<GenericTypeParamDecl>())
            {
                if (!first) sb << ", ";
                first = false;

                sb << getText(genericTypeParam.getName());
            }
            else if (auto genericValParam = paramDeclRef.as<GenericValueParamDecl>())
            {
                if (!first) sb << ", ";
                first = false;

                sb << getText(genericValParam.getName());
                sb << ":";
                addType(getType(m_astBuilder, genericValParam));
            }
            else
            {
            }
        }
        sb << ">";

        addDeclParams(DeclRef<Decl>(getInner(genericDeclRef), genericDeclRef.substitutions));
    }
    else
    {
    }
}

void ASTPrinter::addDeclKindPrefix(Decl* decl)
{
    if (auto genericDecl = as<GenericDecl>(decl))
    {
        decl = genericDecl->inner;
    }
    if (as<FuncDecl>(decl))
    {
        m_builder << "func ";
    }
}

void ASTPrinter::addDeclResultType(const DeclRef<Decl>& inDeclRef)
{
    DeclRef<Decl> declRef = inDeclRef;
    if (auto genericDeclRef = declRef.as<GenericDecl>())
    {
        declRef = DeclRef<Decl>(getInner(genericDeclRef), genericDeclRef.substitutions);
    }

    if (as<ConstructorDecl>(declRef))
    {
    }
    else if (auto callableDeclRef = declRef.as<CallableDecl>())
    {
        m_builder << " -> ";
        addType(getResultType(m_astBuilder, callableDeclRef));
    }
}

/* static */void ASTPrinter::addDeclSignature(const DeclRef<Decl>& declRef)
{
    addDeclKindPrefix(declRef.getDecl());
    addDeclPath(declRef);
    addDeclParams(declRef);
    addDeclResultType(declRef);
}

/* static */String ASTPrinter::getDeclSignatureString(DeclRef<Decl> declRef, ASTBuilder* astBuilder)
{
    ASTPrinter astPrinter(astBuilder);
    astPrinter.addDeclSignature(declRef);
    return astPrinter.getString();
}

/* static */String ASTPrinter::getDeclSignatureString(const LookupResultItem& item, ASTBuilder* astBuilder)
{
    return getDeclSignatureString(item.declRef, astBuilder);
}

} // namespace Slang
