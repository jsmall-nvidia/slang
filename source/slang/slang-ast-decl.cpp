// slang-ast-decl.cpp
#include "slang-ast-builder.h"
#include <assert.h>

namespace Slang {

const TypeExp& TypeConstraintDecl::getSup() const
{
    SLANG_AST_NODE_CONST_VIRTUAL_CALL(TypeConstraintDecl, getSup, ())
}

const TypeExp& TypeConstraintDecl::_getSupOverride() const
{
    SLANG_ASSERT(!"TypeConstraintDecl::_getSupOverride not overridden");
    return TypeExp::empty;
}


} // namespace Slang
