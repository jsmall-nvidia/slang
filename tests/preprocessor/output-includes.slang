//DIAGNOSTIC_TEST:SIMPLE:-output-includes

int foo() { return 0; }

#include "include-a.slang.h"
#include "../preprocessor/include-a.slang.h"
#include "./include-a.slang.h"

int baz() { return bar(); }