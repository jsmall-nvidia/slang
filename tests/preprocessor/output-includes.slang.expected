result code = -1
standard error = {
(0): note: include 'tests/preprocessor/output-include.slang'
(0): note: include '  tests/preprocessor/include-a.slang.h'
(0): note: include '  tests/preprocessor/include-a.slang.h'
(0): note: include '  tests/preprocessor/include-a.slang.h'
(0): note: include '  tests/preprocessor/include/pragma-once-c.h'
tests/preprocessor/include-a.slang.h(3): error 30201: function 'bar' already has a body
tests/preprocessor/include-a.slang.h(3): note: see previous definition of 'bar'
tests/preprocessor/include-a.slang.h(3): error 30201: function 'bar' already has a body
tests/preprocessor/include-a.slang.h(3): note: see previous definition of 'bar'
}
standard output = {
}
